src/fsharp/FSharp.Core/async.fs

```diff
@@ -111,11 +105,31 @@ namespace Microsoft.FSharp.Control
                     Trampoline.thisThreadHasTrampoline <- true
                     true
             try
-                loop firstAction
+                let mutable keepGoing = true
+                let mutable action = firstAction
+                while keepGoing do 
+                    try 
+                        action() |> unfake
+                        match storedCont with
+                        | None ->
+                            keepGoing <- false
+                        | Some cont -> 
+                            storedCont <- None
+                            action <- cont
+                    // Let the exception propagate all the way to the trampoline to get a full .StackTrace entry
+                    with exn -> 
+                        match storedExnCont with
+                        | None ->
+                            reraise()
+                        | Some econt -> 
+                            storedExnCont <- None
+                            let edi = ExceptionDispatchInfo.RestoreOrCapture exn
+                            action <- (fun () -> econt edi)
+                        
             finally
                 if thisIsTopTrampoline then
                     Trampoline.thisThreadHasTrampoline <- false
-            FakeUnit
+            Unchecked.defaultof<AsyncReturn>
```

Comment:
This is essentially `null`, correct? Should we add `AllowNull` attribute and use `null`? And add a comment that only `null` is used as value for this type?

Reply:
This is just a question, I guess the current solution is fine - just a bit unexpected when looking at the type